File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
private:
    int binary_search(vector<int>& nums, int target, bool flag) {
        int left = 0, right = nums.size() - 1;
        int result = -1;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (nums[mid] == target) {
                result = mid;
                flag ? right = mid - 1 : left = mid + 1;
            }
            else if (nums[mid] < target) left = mid + 1;
            else right = mid - 1;
        }
        return result;
    }
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        return {binary_search(nums, target, true), binary_search(nums, target, false)};
    }
};

File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        
        numsLen = len(nums)
        
        def binarySearch() -> int:
            
            left = 0
            right = numsLen - 1
            
            while left <= right:
                
                mid = int((left + right) / 2)
                if nums[mid] == target: return mid
                if (nums[mid] < target): left = mid + 1
                else: right = mid - 1
            
            return -1
        
        idx = binarySearch()
        
        if idx < 0: return [-1,-1]
        
        start = idx
        end = idx
        
        while start > 0 and nums[start] == target:
            start -= 1
        
        if start > 0 or nums[start] != target: start += 1
            
        while end < numsLen and nums[end] == target:
            end += 1
        
        end -= 1
            
        return [start, end]
            
            